### ISSUE TITLE
5 or more cameras connect issue

### ISSUE THREAD

**Issue by MehmetOKUYAR** [basler/pypylon#497]
![image](https://user-images.githubusercontent.com/58508173/169837362-65568fd2-a224-487a-ab2d-6dbe84dcc2a1.png)

I get this error when I try to use 5 or more usb basler cameras

`_genicam.RuntimeException: WaitObject duplicate failed (0): Too many open files. Reached open files limit: To change the limit permanently on a ubuntu system, edit /etc/security/limits.conf : RuntimeException thrown (file 'WaitObject_Unix.cpp', line 76)
`

and also ı get this error 

`_genicam.RuntimeException: PrepareGrab (StartStreaming) failed for device '2676:ba02:4:4:5'. Error: 'Insufficient system resources exist to complete the API.' : RuntimeException thrown (file 'PylonUsbStream.cpp', line 275)
`

how can ı solve ? is there any known solution ?

**Comment by SMA2016a** [comment]
Try to increase the number of file descriptor on your system

ulimit -n 4096

**Comment by MehmetOKUYAR** [comment]
actually, before I tried this code, I couldn't open more than 4 cameras in the pylon Viewer interface. After trying this code, I managed to run 4 or more cameras in the pylon Viewer interface, but in the python code I wrote, I still get an error when running more than 4 cameras.

**Comment by SMA2016a** [comment]
![grafik](https://user-images.githubusercontent.com/18481846/170017570-32884380-8085-46bf-9e4a-db295626cb12.png)

**Comment by SMA2016a** [comment]
reboot your system and then remove the following lines.

window_name = f'Camera-{cameraContextValue:03}'
            cv2.imshow(window_name, img)
            k = cv2.waitKey(1)
            if k == 27:
                break

then try to run your code, and let me know, if you can use 5 cameras without any issue

**Comment by MehmetOKUYAR** [comment]
@SMA2016a Yes, ı didnt get an error like that, it worked but ı want to show images. how to is possible?

**Comment by thiesmoeller** [comment]
One error in your code:
you never return the buffers to the buffer queue ( see example on readme.md ).
One way to let pypylon take care of this is the following pattern
 
```python
while cam.IsGrabbing():
    with cam.RetrieveResult(1000) as res:
        if res.GrabSucceeded():
            <do something with _res_ >


```

Here the context manager of the `with` statement will take care to release `res` in any case, when the context is left.

If you never return buffers to the grab-loop you either run out of buffers or your operating system can't handle so many open data structures ... what ever comes first

**Comment by SMA2016a** [comment]
@MehmetOKUYAR  in your code, it is  missing :   grabResult.Release()

if grabResult.GrabSucceeded():
        # Access the image data
        image = converter.Convert(grabResult)
        img = image.GetArray()
        cv2.namedWindow('title', cv2.WINDOW_NORMAL)
        cv2.imshow('title', img)
        k = cv2.waitKey(1)
        if k == 27:
            break
    grabResult.Release()

**Comment by SMA2016a** [comment]
@MehmetOKUYAR   OK, then close this thread

**Comment by MehmetOKUYAR** [comment]
> Try to increase the number of file descriptor on your system
> 
> echo 1000 > /sys/module/usbcore/parameters/usbfs_memory_mb
> 
> ulimit -n 4096
> 
> let me know which one helps you

it worked with these
